src/io/utils.rs

```
@@ -200,34 +204,118 @@ where
+pub(crate) fn migrate_deprecated_spendable_outputs<K: KVStore + Sync + Send, L: Deref>(
+	sweeper: Arc<Sweeper<K>>, kv_store: Arc<K>, logger: L,
+) -> Result<(), std::io::Error>
 where
 	L::Target: Logger,
 {
-	let mut res = Vec::new();
+	let best_block = sweeper.current_best_block();
 
 	for stored_key in kv_store.list(
-		SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
-		SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+		DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+		DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
 	)? {
 		let mut reader = Cursor::new(kv_store.read(
-			SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
-			SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+			DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+			DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
 			&stored_key,
 		)?);
-		let output = SpendableOutputInfo::read(&mut reader).map_err(|e| {
+		let output = DeprecatedSpendableOutputInfo::read(&mut reader).map_err(|e| {
 			log_error!(logger, "Failed to deserialize SpendableOutputInfo: {}", e);
 			std::io::Error::new(
 				std::io::ErrorKind::InvalidData,
 				"Failed to deserialize SpendableOutputInfo",
 			)
 		})?;
-		res.push(output);
+		let descriptors = vec![output.descriptor.clone()];
+		let spend_delay = Some(best_block.height + 2);
+		sweeper.track_spendable_outputs(descriptors, output.channel_id, true, spend_delay);
+		if let Some(tracked_spendable_output) =
+			sweeper.tracked_spendable_outputs().iter().find(|o| o.descriptor == output.descriptor)
+		{
+			match tracked_spendable_output.status {
+				OutputSpendStatus::PendingInitialBroadcast { delayed_until_height } => {
+					if delayed_until_height == spend_delay {
+						kv_store.remove(
+							DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+							DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+							&stored_key,
+							false,
+						)?;
```

Should we persist the `Sweeper` before removing these?

---

Oh, never mind. It looks like `track_spendable_outputs` does just that.

---

Though it will only log but otherwise fail silently. Is this a concern?

---

Feel free to squash. We can update this in a follow-up as soon as a new RC is available.